### PR TITLE
Attraction Support

### PR DESCRIPTION
**Pull request by Jetz72**
First draft of Attractions. Still has a few hurdles to clear but you can add an Attraction section to a deck and play them in game (with deck conformance disabled since there are only two so far). 

The weirdest new component to all this, and one I was looking for feedback on, is something I'm calling "functional variants" - a property used to reconcile different prints with the same card name having different behavior (namely, the lights).
The script for an Attraction looks like this:
```
Name:Information Booth
ManaCost:no cost
Types:Artifact Attraction
Variant:A:Lights:2 6
Variant:B:Lights:3 6
Variant:C:Lights:4 6
Variant:D:Lights:5 6
K:Visit:TrigDraw
SVar:TrigDraw:DB$ Draw | SpellDescription$ Draw a card.
Oracle:Visit — Draw a card.
```
The lines beginning with `Variant:Name` match to the same name specified in an optional additional parameter in the edition file:
```
218a U Information Booth @Gaboleps $A
218b U Information Booth @Gaboleps $B
218c U Information Booth @Gaboleps $C
218d U Information Booth @Gaboleps $D
```
So far it's only been tested with Lights, but it's designed to work with any property. The intent was that it would also enable implementations for some variants of other Un-cards, such as Everythingamajig and Garbage Elemental. There are still some parts of the implementation that need smoothing out, but it works as a proof-of-concept for now, and some input would be nice before spending more time polishing it up.

I think most of the rest is relatively straightforward, with the Attraction deck largely precedented by Planechase and Scheme decks. There are still some things left to do, which is why I'm putting this up as a draft for now. My list includes:
* Prize keyword
* TriggerClaimPrize
* Mobile Deck builder support
* Adventure deck builder support?
* Figure out the merge move logic - If you animate one, mutate something onto it, and then it leaves the battlefield, where does it end up? Does it matter if a token or commander is involved?
* "number of attractions you visited this turn" - It's tracked but can't yet reference it in script. I'm unclear whether visiting the same attraction twice in a turn counts twice for this.
* Deck generation rules - attraction synergy + generating an attraction deck if needed
* AI logic
* Detail description for Visit keyword. Currently shows Visit:TrigScry
* Attraction roll result dialog has an extra "Result:" in its text

Nice-to-haves:
* Astrotorium card backs
* Light icons in the card details instead of numbers
* Reorganize RollDiceEffect.rollDiceForPlayer to remove dependency on SpellAbility - Right now rollToVisitYourAttractions() copypastes most of rollDiceForPlayer
* Move functional variant properties to a different ICardFace implementation

### CODE REVIEW

File: forge-core/src/main/java/forge/card/CardRules.java [hunk 1]
```diff
@@ -546,9 +563,7 @@ public final void parseLine(final String line) {
                     } else if ("DeckHas".equals(key)) {
                         has = new DeckHints(value);
                     } else if ("Defense".equals(key)) {
-                        this.faces[this.curFace].setDefense(value);
-                    } else if ("Draft".equals(key)) {
-                        this.faces[this.curFace].addDraftAction(value);
```
Comment (tool4ever) [inline forge-core/src/main/java/forge/card/CardRules.java:551]: bad rebase?

File: forge-game/src/main/java/forge/game/phase/PhaseHandler.java [hunk 2]
```diff
@@ -281,12 +281,16 @@ private final void onPhaseBegin() {
...
                         GameEntityCounterTable table = new GameEntityCounterTable();
-                        // all Saga get Lore counter at the begin of pre combat
+                        // all Sagas get a Lore counter at the beginning of pre combat
                         for (Card c : playerTurn.getCardsIn(ZoneType.Battlefield)) {
                             if (c.isSaga()) {
                                 c.addCounter(CounterEnumType.LORE, 1, playerTurn, table);
                             }
                         }
+                        // roll for attractions if we have any
+                        if (CardLists.count(playerTurn.getCardsIn(ZoneType.Battlefield), Presets.ATTRACTIONS) > 0) {
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/phase/PhaseHandler.java:291]: `Iterables.any` is better for performance I think

File: forge-core/src/main/java/forge/card/CardRules.java (hunk 1 above)
Comment (Jetz72) [reply forge-core/src/main/java/forge/card/CardRules.java:551]: Didn't realize at 3 AM that the IDE's merge tool was showing me the whole file and not just a single change. Fixed now.

File: forge-game/src/main/java/forge/game/phase/PhaseHandler.java (hunk 2 above)
Comment (Jetz72) [reply forge-game/src/main/java/forge/game/phase/PhaseHandler.java]: Also reads better. Done.

File: forge-game/src/main/java/forge/game/player/Player.java [hunk 3]
```diff
@@ -3812,4 +3832,81 @@ public int getCommittedCrimeThisTurn() {
...
     }
+
+    public void visitAttractions(int light) {
+        CardCollection attractions = CardLists.filter(getCardsIn(ZoneType.Battlefield), CardPredicates.isAttractionWithLight(light));
+        if(attractions.isEmpty())
+            return;
+        for (Card c : attractions) {
+            incrementAttractionsVisitedThisTurn();
+
+            final Map<AbilityKey, Object> runParams = AbilityKey.newMap();
```
[the remark this answers is not in the document]
Comment (Jetz72) [reply forge-game/src/main/java/forge/game/player/Player.java]: Now if only I could remember where I was sampling this from, I could get that one too...

**Comment by Jetz72** [comment]
Moving from draft PR to full PR. 

Features added:
Attraction support
7 Attraction support cards
13 Attraction cards
Functional Variant support
3 cards that use functional variants, two with one variant implemented, one with two variants.
Overhaul for Deck Editor on mobile - now supports extra deck sections for variant cards.
-TabPageScreen now supports adding new tabs after initialization.

Things not yet done, investigated, and/or tested:
Draft/Limited play
Adventure Mode Deck Editor
Deck generation, generating an Attraction deck if needed and adding Attraction synergy cards to the main deck
Images on downloads.cardforge.org
Some better icons for deck sections on Mobile

Things that'd be nice to do at some point down the line:
Light icons in the card details instead of numbers
Astrotorium card backs
A way to view the Attraction Deck and Junkyard zones

File: forge-core/src/main/java/forge/deck/DeckRecognizer.java [hunk 4]
```diff
@@ -760,6 +762,9 @@ private DeckSection getTokenSection(String deckSec, DeckSection currentDeckSecti
         // is not supported, but other possibilities exist (e.g. Commander card in Constructed
         // could potentially go in Main)
         DeckSection matchedSection = DeckSection.matchingSection(card);
+        // If it's a commander candidate, put it there.
+        if (matchedSection == DeckSection.Main && this.isAllowed(DeckSection.Commander) && DeckSection.Commander.validate(card))
```
[the remark this answers is not in the document]
Comment (Jetz72) [reply forge-core/src/main/java/forge/deck/DeckRecognizer.java:766]: I edited `DeckSection.matchingSection` to make it more generally usable, by removing the test that put all legendary creatures in the commander section. While investigating to see if this would break any current uses of it, I stumbled upon DeckRecognizerTest.java and ran it. There was a corner case test called `testCornerCaseWhenThereIsNoCurrentSectionAndMatchedSectionIsNotAllowedButCouldMatchMain()` (might be the longest symbol name I've ever seen by the way), where it was apparently the intended behavior that `DeckSection.matchingSection` would return `DeckSection.Commander` for commander candidates.

That was the only case I could find where it seemed to matter, and since there was a test for it specifically, I assumed it was intentional. So I shifted the responsibility of choosing the Commander section outward to that `getTokenSection` method, in the form of the lines there.

File: forge-game/src/main/java/forge/game/GameAction.java [hunk 5]
```diff
@@ -155,6 +155,16 @@ private Card changeZone(final Zone zoneFrom, Zone zoneTo, final Card c, Integer
             }
         }
 
+        //717.6. If a card with an Astrotorium card back would be put into a zone other than the battlefield, exile,
+        //or the command zone from anywhere, instead its owner puts it into the junkyard.
+        if (c.isAttractionCard() && !toBattlefield && !zoneTo.is(ZoneType.AttractionDeck)
+                && !zoneTo.is(ZoneType.Junkyard) && !zoneTo.is(ZoneType.Exile) && !zoneTo.is(ZoneType.Command)) {
+            //This should technically be a replacement effect, but with the "can apply more than once to the same event"
+            //clause, this seems sufficient for now.
+            //TODO: Figure out what on earth happens if you animate an attraction, mutate a creature/commander/token onto it, and it dies...
```
[the remark this answers is not in the document]
Comment (tehdiplomat) [reply forge-game/src/main/java/forge/game/GameAction.java:164]: Probably ok for now, but we should keep an eye on this if reports come in.

**Review by tehdiplomat — DISMISSED** [review]

**Review by tehdiplomat — APPROVED** [review]
